### ISSUE TITLE
Cheap moves and exchanges are missing from opening move evaluation

### ISSUE THREAD

**Issue by laeuter** [quackle/quackle#111]
After simulation with another program and a German word list there are signs that optimal play in the first move involves around
* 40 % exchanges
* 7 % moves of 2-letter words.

Quackle's Championship Player doesn't achieve these numbers (using the German word list) -- doing only 4 % exchanges and 2 % 2-letter words --, apparently by filtering out cheap plays and not considering exchanges when there are enough valid words available. Apparently massaging the rack for a bingo in the next move is an item in optimal play.

If Quackle evaluates exchanges, it finds some of the strong ones, but misses many.

Example: (on blank board)
Rack = NLEOKAS
Quackle: 8D KOSEN (V=31.5106, BW=55.3138%), 8D KLONS (V=31.7303, BW=55.3062%), 8D ONKEL (V=31.4106, BW=55.293%), ... no exchange at all
My simulation: 8H OS (BW=61.148), 8F LOK (BW=60.869), (EXCH) O (BW=60.613), 8F KLO (BW=60.569), 8G SO (BW=60.409), ..., 8D KOSEN (BW=58.113), ..., 8D ONKEL (BW=57.633), 8D KLONS (BW=55.238)

There are four 2-letter moves (with 6 points each), nine 3-letter moves, four 4-letter moves and two exchanges evaluating better than 8D KOSEN with 26 points, but the Quackle candidates contain none of them. A variation of 3 % in bogowin is known to be normal, but the elision of _all_ cheap moves appears to be a regression.

Is it possible to configure evaluation so that more of the cheap plays, and especially exchanges, are evaluated, at least in the first few moves? Today's computers are so fast, what is wrong with evaluating _all_ valid moves?

**Comment by jfultz** [comment]
I agree with you that we can go deeper into move evaluation.  The Championship Player statically evaluates a certain number of moves, and then further evaluates a subset via simulation. 
 Threading the Championship evaluator and upping the number of candidates it checks is on my short list of things to do, issue #93.

But the simulations, while getting faster, are still not fast enough to sufficiently consider all candidate moves throughout the game and play a good, 25 minute game.  Even on an opening move, which typically has fewer options than a mid-game move, there may be hundreds of plays (RETAINS in NWL18 has 277 subanagrams) to analyze even before exchanges are considered.  The Championship Player uses a minimum of 20 sims on any play it chooses to sim, so that's easily into thousands of sims.

We could choose to statically evaluate more moves, but it's not clear to me that our current static evaluation is going to lead to the kinds of improvements you're seeking.

So, the progress forward is to resolve issue #93 first, which will get us faster and deeper into the candidates.  Maybe not deep enough, in which case I'll investigate a bit more how we cull for static and sim evaluation with your examples in mind.

Incidentally, you claim that there's a regression.  It would help if you were more specific.  Are you claiming that there is a regression specifically in the evaluation of the opening rack NLEOKAS, and if so, what version are you comparing to?

**Comment by laeuter** [comment]
Sorry for the 'regression' wording. I didn't want to compare Quackle to previous versions, but to simulation results from other programs. It's not the handling of just this example rack, but it's the general cut-off of cheap plays in opening moves, that I wanted to point at.

Elise has the same problem, but a different cut-off point for cheap moves. And a much weaker static exchange selector than Quackle. But *when* it considers cheap plays, it puts them in the same places above the high-scoring plays as my simulation. It was the cross comparison of the 3 programs that made me investigate the cheap opening moves. With Elise also pointing to this hole, it shouldn't be a problem of my program alone.

You are right that some racks (frequent letters plus joker(s)) allow for hundreds of word candidates. Plus 128 exchange moves. Most of them fall off after a few (e.g. 80) sim racks. My observation was that with cutting off cheap plays Quackle loses considerable ground right at the beginning of the game. I feel that is like defeating the purpose of Quackle.

FYI when simulating 20,000 random game openings, bingo plays lost just once to a none-bingo (STORY just before TROYE(R)S). So cutting off should be acceptable when a bingo is on the rack. I only gave one example in the issue, but I can give thousands, the ratio of exchanges being 39 % and winning 2- or 3-letter words 22 %.

N. B.: I am not so sure about the number of options when comparing opening moves to mid-game moves. It might be true, but mid-game moves somehow simulate much faster than opening moves or the complete end game. The end game is wide and can be deep, so it should take long. Perhaps mid-game is faster due to the restrictions of the letters already present on the board.
